### PR TITLE
Show the scm status request response in the workflow run

### PR DESCRIPTION
**Pull request by danidoni**
While running a workflow run, we may have errors when sending the statuses to the SCM.
When that happens we save the response into a new `ScmStatusReport` object so we can debug it later.

**This is how it looks like**
Screenshot from 2022-05-30 12-52-17
Screenshot from 2022-05-30 13-02-08

### CODE REVIEW

File: src/api/app/models/scm_status_report.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+class ScmStatusReport < ApplicationRecord
+  belongs_to :workflow_run
+end
+
+# == Schema Information
+#
+# Table name: scm_status_reports
+#
+#  id              :integer          not null, primary key
+#  response_body   :text(4294967295)
```
Comment (hennevogel) [inline src/api/app/models/scm_status_report.rb:10]: I think you also want to record what you send (request), analog to the webhooks. Like whatever is happening in `SCMStatusReporter.call`

Comment (danidoni) [reply src/api/app/models/scm_status_report.rb:10]: You mean the request payload and it's headers?

File: src/api/app/models/workflow_run.rb [hunk 2]
```diff
@@ -27,10 +28,25 @@ class WorkflowRun < ApplicationRecord
...
   }
 
-  def update_to_fail(message)
+  # Marks the workflow run as failed and records the relevant debug information in response_body
+  def update_as_failed(message)
     update(response_body: message, status: 'fail')
   end
 
+  # Stores debug info to help figure out what went wrong when trying to save a Status in the SCM.
+  def report_failure_to_scm(message, options)
```
Comment (krauselukas) [inline src/api/app/models/workflow_run.rb:37]: @danidoni I find the naming a bit confusing, I was expecting that this method is reporting something to the scm, not saving the report in OBS.

Comment (krauselukas) [reply src/api/app/models/workflow_run.rb:37]: maybe something like `save_failed_scm_status_report`

File: src/api/app/models/token/errors.rb [hunk 3]
```diff
@@ -7,6 +7,12 @@ class NoReleaseTargetFound < APIError
 
   class NonExistentRepository < APIError
     setup 404
+
+    def initialize(error, request_parameters)
+      super(error)
+
+      @request_parameters = request_parameters
```
[the remark this answers is not in the document]
Comment (danidoni) [reply src/api/app/models/token/errors.rb]: True, removing

File: src/api/app/models/workflow_run.rb (hunk 2 above)
Comment (danidoni) [reply src/api/app/models/workflow_run.rb]: Yes, perfect! Changed

**Comment by krauselukas** [comment]
@danidoni Generally it looks good to me, the only thing I would improve (but probably in an follow up) is how we handle successful reports to the scm in the `Reports to the SCM` tab. If its successful we show that there were no status reports. Thats confusing, there were no failing ones. Maybe that would already be an improvement for now. Instead of showing `No status reports` something like `No failed status reports`.

File: src/api/app/components/workflow_run_detail_component.html.haml [hunk 4]
```diff
@@ -22,6 +26,22 @@
...
+            %h6 Response body
+            %pre.border.p-2#status-report-response-body
+              = status_report.response_body
+            %h6 Request parameters
+            %pre.border.p-2#status-report-request-parameters
+              = status_report.pretty_request_parameters
+            %small.text-nowrap{ title: l(status_report.created_at.utc) }
+              Sent #{time_ago_in_words(status_report.created_at)} ago
+    - else
+      No status reports
```
Comment (krauselukas) [inline src/api/app/components/workflow_run_detail_component.html.haml:44]:
```suggestion
      No failed status reports
```

Comment (krauselukas) [reply src/api/app/components/workflow_run_detail_component.html.haml:44]: thats what I meant in my previous comment

**Comment by krauselukas** [comment]
the codeclimate stuff we can ignore for now if you want and maybe create a card for the refactoring :)

**Comment by danidoni** [comment]
Code climate issues will be dealt on #12646

**Review by rubhanazeem — APPROVED** [review]

File: src/api/app/components/workflow_run_detail_component.html.haml [hunk 5]
```diff
@@ -22,6 +26,24 @@
...
       = pretty_request_payload
+  .tab-pane.fade{ id: "scm-reports-tab-content#{id}", role: 'tabpanel',
+                  aria: { labelledby: "scm-reports-tab#{id}" } }
+    - if status_reports.any?
+      .list-group.list-group-flush
+        - status_reports.each do |status_report|
+          .list-group-item
+            %h6 Response body
+            %p
+              %span.badge{ class: status_report.status_class }= status_report.status.humanize
```
Comment (krauselukas) [inline src/api/app/components/workflow_run_detail_component.html.haml:37]: I tested with GitHub and had cases where `status_report.status` is `nil`. This explodes when calling `humanize` on it

File: src/api/app/models/scm_status_report.rb [hunk 6]
```diff
@@ -0,0 +1,43 @@
+# This class represents the communication that took place between OBS and the
+# SCM to render each one of the check marks.
+#
+# Those check marks appear as green, orange and red ticks at the bottom of a
+# PR or next to a commit, on GitHub.
+#
+# The details of the request and the response will be shown in the
+# Reports to the SCM tab in Workflow Runs show page.
+class ScmStatusReport < ApplicationRecord
```
Comment (krauselukas) [inline src/api/app/models/scm_status_report.rb:9]:
```suggestion
class SCMStatusReport < ApplicationRecord
```

Comment (krauselukas) [reply src/api/app/models/scm_status_report.rb:9]: just realized we write it upercase everywhere else, you need to set this in the `zeitwerk.rb` file

**Review by krauselukas — CHANGES_REQUESTED** [review]

File: src/api/app/components/workflow_run_detail_component.html.haml (hunk 5 above)
Comment (krauselukas) [reply src/api/app/components/workflow_run_detail_component.html.haml:37]: Seems the default value on the `status` column doesn't have an effect

Comment (krauselukas) [reply src/api/app/components/workflow_run_detail_component.html.haml]: okay, havent realized that the migration changed meanwhile... my DB was outdated

File: src/api/app/models/scm_status_report.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
+# This class represents the communication that took place between OBS and the
+# SCM to render each one of the check marks.
+#
+# Those check marks appear as green, orange and red ticks at the bottom of a
+# PR or next to a commit, on GitHub.
+#
+# The details of the request and the response will be shown in the
+# Reports to the SCM tab in Workflow Runs show page.
+class SCMStatusReport < ApplicationRecord
```
Comment (krauselukas) [inline src/api/app/models/scm_status_report.rb:9]:
```suggestion
class ScmStatusReport < ApplicationRecord
```

Comment (krauselukas) [reply src/api/app/models/scm_status_report.rb:9]: Its not working for me with uppercase, I guess this is not possible with Models...

Comment (krauselukas) [reply src/api/app/models/scm_status_report.rb:9]: I get this 
```
[f245a8f7-842b-497f-946a-e7d67df5ce51] ActionView::Template::Error (uninitialized constant WorkflowRun::ScmStatusReport

            raise NameError.new("uninitialized constant #{candidates.first}", candidates.first)
            ^^^^^):
[f245a8f7-842b-497f-946a-e7d67df5ce51]     2: .card
[f245a8f7-842b-497f-946a-e7d67df5ce51]     3:   .card-body
[f245a8f7-842b-497f-946a-e7d67df5ce51]     4:     = render(WorkflowRunHeaderComponent.new(workflow_run: @workflow_run))
[f245a8f7-842b-497f-946a-e7d67df5ce51]     5:     = render(WorkflowRunDetailComponent.new(workflow_run: @workflow_run))
[f245a8f7-842b-497f-946a-e7d67df5ce51]   
[f245a8f7-842b-497f-946a-e7d67df5ce51] app/components/workflow_run_detail_component.rb:15:in `initialize'
[f245a8f7-842b-497f-946a-e7d67df5ce51] app/views/webui/workflow_runs/show.html.haml:5:in `new'
[f245a8f7-842b-497f-946a-e7d67df5ce51] app/views/webui/workflow_runs/show.html.haml:5
[f245a8f7-842b-497f-946a-e7d67df5ce51] config/initializers/wrap_parameters.rb:37:in `call'
```

**Review by krauselukas — CHANGES_REQUESTED** [review]

File: src/api/app/models/scm_status_report.rb (hunk 7 above)
Comment (danidoni) [reply src/api/app/models/scm_status_report.rb]: The same on my side, I'll revert this inflection and fix it later

**Pull request merged**